src/imex/transfer.rs

```diff
@@ -0,0 +1,539 @@
+    }
+
+    /// Supervises the sendme [`Provider`], terminating it when needed.
+    ///
+    /// This will watch the provider and terminate it when:
+    ///
+    /// - A transfer is completed, successful or unsuccessful.
+    /// - An event could not be observed to protect against not knowing of a completed event.
+    /// - The ongoing process is cancelled.
+    ///
+    /// The *cancel_token* is the handle for the ongoing process mutex, when this completes
+    /// we must cancel this operation.
+    async fn watch_provider(
+        context: Context,
+        mut provider: Provider,
+        cancel_token: Receiver<()>,
+        dbfile: PathBuf,
+    ) -> Result<()> {
+        let mut events = provider.subscribe();
+        let res = loop {
+            tokio::select! {
+                biased;
+                res = &mut provider => {
+                    break res.context("BackupSender failed");
+                },
+                maybe_event = events.recv() => {
+                    match maybe_event {
+                        Ok(event) => {
+                            match event {
+                                Event::ClientConnected { ..} => {
+                                    context.emit_event(SendProgress::ClientConnected.into());
+                                }
+                                Event::RequestReceived { .. } => {
+                                    context.emit_event(SendProgress::TransferStarted.into());
+                                }
+                                Event::TransferCompleted { .. } => {
+                                    context.emit_event(SendProgress::TransferFinished.into());
+                                    provider.shutdown();
+                                }
+                                Event::TransferAborted { .. } => {
+                                    provider.shutdown();
+                                    break Err(anyhow!("BackupSender transfer aborted"));
+                                }
+                            }
+                        }
+                        Err(broadcast::error::RecvError::Closed) => {
+                            // We should never see this, provider.join() should complete
+                            // first.
+                        }
+                        Err(broadcast::error::RecvError::Lagged(_)) => {
+                            // We really shouldn't be lagging, if we did we may have missed
+                            // a completion event.
+                            provider.shutdown();
+                            break Err(anyhow!("Missed events from BackupSender"));
+                        }
+                    }
+                },
+                _ = cancel_token.recv() => {
+                    provider.shutdown();
+                    break Err(anyhow!("BackupSender cancelled"));
+                },
+            }
+        };
+        if let Err(err) = fs::remove_file(&dbfile).await {
```

Comment:
Because there is no `async drop` it is pretty much impossible to reliably use guard object for async code execution unfortunately.

Reply:
You can, but then you have no idea when that task will be finished

Reply:
But why should i care? If someone needs to wait for it, a Mutex, JoinHandle or smth else can be used

Reply:
I guess in this case there's no need to know when the file is deleted, so I've moved this to a drop guard.  It even fixes a bug in case of error in `prepare_inner()`!